pkg/operator/controller/singleton.go

```diff
@@ -105,23 +92,90 @@ func (s *Singleton) Start(ctx context.Context) error {
 	measureDuration() // Observe the length of time between the function creation and now
 
 	switch {
 	case err != nil:
-		s.metrics.reconcileErrors.Inc()
+		reconcileErrors.WithLabelValues(s.metricName()).Inc()
+		reconcileTotal.WithLabelValues(s.metricName(), labelError).Inc()
 		logging.FromContext(ctx).Error(err)
 		return s.rateLimiter.When(singletonRequest)
 	case res.Requeue:
+		reconcileTotal.WithLabelValues(s.metricName(), labelRequeue).Inc()
 		return s.rateLimiter.When(singletonRequest)
 	default:
 		s.rateLimiter.Forget(singletonRequest)
-		return lo.Ternary(res.RequeueAfter > 0, res.RequeueAfter, time.Duration(0))
+		switch {
+		case res.RequeueAfter > 0:
+			reconcileTotal.WithLabelValues(s.metricName(), labelRequeueAfter).Inc()
+			return res.RequeueAfter
+		default:
+			reconcileTotal.WithLabelValues(s.metricName(), labelSuccess).Inc()
+			return time.Duration(0)
+		}
 	}
 }
 
 func (s *Singleton) NeedLeaderElection() bool {
 	return true
 }
+
+func init() {
+	mergeMetrics()
+}
+
+const (
+	labelError        = "error"
+	labelRequeueAfter = "requeue_after"
+	labelRequeue      = "requeue"
+	labelSuccess      = "success"
+)
+
+// Metrics below are copied metrics fired by controller-runtime in its /internal package. This is leveraged
+// so that we can fire to the same namespace as users expect other controller-runtime metrics to be fired
+// https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/internal/controller/metrics/metrics.go
+var (
+	reconcileTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "controller_runtime_reconcile_total",
+		Help: "Total number of reconciliations per controller",
+	}, []string{"controller", "result"})
+	reconcileDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "controller_runtime_reconcile_time_seconds",
+		Help:    "Length of time per reconciliation per controller",
+		Buckets: metrics.DurationBuckets(),
+	}, []string{"controller"})
+	reconcileErrors = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "controller_runtime_reconcile_errors_total",
+		Help: "Total number of reconciliation errors per controller",
+	}, []string{"controller"})
+	workerCount = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "controller_runtime_max_concurrent_reconciles",
+		Help: "Maximum number of concurrent reconciles per controller",
+	}, []string{"controller"})
+	activeWorkers = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "controller_runtime_active_workers",
+		Help: "Number of currently used workers per controller",
+	}, []string{"controller"})
+)
+
+// mergeMetrics merges the singletonMetrics with metrics already registered in the controller-runtime metrics registry
+// https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/internal/controller/metrics/metrics.go
+// We know that all these metrics should be registered by controller-runtime so we should switch over
+func mergeMetrics() {
```

Comment:
Though this seems slightly hacky, this is a supported method for switching over to a metric that has already been registered, so we can leverage this to switch over the controller-runtime metrics when firing for the singleton controller

Reply:
See: https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L205